src/Firestore/FirestoreClient.php

```diff
@@ -0,0 +1,455 @@
+     * execution. Thrown exceptions will prevent commit and trigger a rollback,
+     * and will bubble up to your level to be handled in whatever fashion is
+     * appropriate.
+     *
+     * Example:
+     * ```
+     * $transferAmount = 500.00;
+     * $from = $firestore->document('users/john');
+     * $to = $firestore->document('users/dave');
+     *
+     * $firestore->runTransaction(function (Transaction $t) use ($from, $to, $transferAmount) {
+     *     $fromSnapshot = $t->snapshot($from);
+     *     $toSnapshot = $t->snapshot($to);
+     *
+     *     $fromNewBalance = $fromSnapshot['balance'] - $transferAmount;
+     *     $toNewBalance = $toSnapshot['balance'] + $transferAmount;
+     *
+     *     // If the transaction cannot be completed, throwing any exception
+     *     // will trigger a rollback operation.
+     *     if ($fromNewBalance < 0) {
+     *         throw new \Exception('User 1 has insufficient funds!');
+     *     }
+     *
+     *     $t->update($from, [
+     *         'balance' => $fromNewBalance
+     *     ])->update($to, [
+     *         'balance' => $toNewBalance
+     *     ]);
+     * });
+     * ```
+     *
+     * @param callable $callable A callable function, allowing atomic operations
+     *        against the Firestore API. Function signature:
+     *        `function (Transaction $t, bool $isRetry)`.
+     * @param array $options {
+     *     Configuration Options.
+     *
+     *     @param array $begin Configuration options for BeginTransaction.
+     *     @param array $commit Configuration options for Commit.
+     *     @param array $rollback Configuration options for rollback.
+     *     @param int $maxRetries The maximum number of times to retry failures.
+     *            **Defaults to** `5`.
+     * }
+     * @return array
+     */
+    public function runTransaction(callable $callable, array $options = [])
+    {
+        if ($this->isRunningTransaction) {
+            throw new \RuntimeException('Another transaction is currently in progress in this client.');
+        }
+
+        $options += [
+            'maxRetries' => self::MAX_RETRIES,
+            'begin' => [],
+            'commit' => [],
+            'rollback' => []
+        ];
+
+        $retryableErrors = [
+            AbortedException::class
+        ];
+
+        $retryFn = function (\Exception $e) use ($retryableErrors) {
+            return in_array(get_class($e), $retryableErrors);
+        };
+
+        // Track the Transaction ID outside the retry function.
+        // If the transaction is retried after an abort, the previous transaction
+        // must be provided to the subsequent `beginTransaction` rpc.
+        // It also provides a convenient indication to the user whether the
+        // transaction is retried or not.
+        $transactionId = null;
+
+        $backoff = new ExponentialBackoff($options['maxRetries'], $retryFn);
```

Comment:
> Furthermore, we should retry failed commits.

This refers to non-transactional commits, correct? Should a backoff be used in that case?

Reply:
"Furthermore, we should retry failed commits."

I think I meant to say we should only retry failed commits (and this also only applies to transactional commits). You can't retry non-transactional commits since they are not idempotent.